DistroLauncher/named_mutex.h

```diff
@@ -0,0 +1,219 @@
+    NamedMutexWrapper(std::wstring name, bool lazy_init = false) :
+        mutex_handle(nullptr), mutex_name(L"WSL_" + DistributionInfo::Name + L"_" + name)
+    {
+        if (!lazy_init) {
+            create();
+        }
+    }
+
+    virtual ~NamedMutexWrapper() noexcept
+    {
+        destroy();
+    }
+
+    class Lock
+    {
+      public:
+        constexpr Lock() noexcept : parent_(nullptr), response_(0)
+        {
+        }
+
+        Lock(NamedMutexWrapper& parent) noexcept : parent_(&parent), response_(parent.wait_and_acquire())
+        {
+        }
+
+        Lock(Lock& other) = delete;
+        Lock(Lock&& other) noexcept : Lock()
+        {
+            *this = std::move(other);
+        }
+
+        ~Lock() noexcept
+        {
+            release();
+        }
+
+        void release() noexcept
+        {
+            if (ok()) {
+                parent_->release();
+            }
+            parent_ = nullptr;
+        }
+
+        Lock& operator=(Lock&& other) noexcept
+        {
+            if (this != &other) {
+                parent_ = std::exchange(other.parent_, nullptr);
+                response_ = std::exchange(other.response_, 0);
+            }
+            return *this;
+        }
+
+        bool ok() const noexcept
+        {
+            return parent_ && (response_ == WAIT_OBJECT_0);
+        }
+
+        DWORD why() const noexcept
+        {
+            return parent_ ? 0 : response_;
+        }
+
+        template <typename Callable> Lock&& and_then(Callable&& f)
+        {
+            if (ok()) {
+                safe_execute(std::forward<Callable>(f), [&]() noexcept { release(); });
+            }
+            return std::move(*this);
+        }
+
+        template <typename Callable> Lock&& or_else(Callable&& f)
+        {
+            if (!ok()) {
+                f();
+            }
+            return std::move(*this);
+        }
+
+      private:
+        NamedMutexWrapper* parent_;
+        DWORD response_;
+
+        friend class NamedMutexWrapper;
+    };
+
+    Lock lock() noexcept
+    {
+        return Lock(*this);
+    }
+
+  protected:
+    HANDLE mutex_handle;
+    std::wstring mutex_name;
+
+    DWORD create() noexcept
+    {
+        return MutexAPI::create(mutex_handle, mutex_name.c_str());
+    };
+    DWORD destroy() noexcept
+    {
+        return MutexAPI::destroy(mutex_handle, mutex_name.c_str());
+    };
+    DWORD wait_and_acquire() noexcept
+    {
+        if (!mutex_handle) {
+            DWORD success = create();
+            if (success != 0) {
+                return success;
+            }
+        }
+        return MutexAPI::wait_and_acquire(mutex_handle, mutex_name.c_str());
+    };
+    DWORD release() noexcept
+    {
+        return MutexAPI::release(mutex_handle, mutex_name.c_str());
+    };
+};
+
+template <typename CallableExec, typename CallablePanic>
+void safe_execute(CallableExec&& f, [[maybe_unused]] CallablePanic&& panic)
```

Comment:
Nitpick: panic is not too dramatic? Wouldn't be more appropriate to call it `onError` or `onFailure`?

Reply:
Fair enough 😄